packages/web3-providers-ws/src/index.ts

```diff
@@ -0,0 +1,387 @@
+		else
+			this.wsProviderOptions = {
+				customTimeout: 1000 * 15,
+			};
+
+		if (!this.wsProviderOptions.customTimeout) this.wsProviderOptions.customTimeout = 1000 * 15;
+
+		if (!this.wsProviderOptions.reconnectOptions)
+			this.wsProviderOptions.reconnectOptions = {
+				auto: false,
+				delay: 5000,
+				maxAttempts: 5,
+				onTimeout: false,
+			};
+
+		this.requestQueue = new Map<JsonRpcId, WSRequestItem>();
+		this.sentQueue = new Map<JsonRpcId, WSRequestItem>();
+
+		this.providerStatus = 'disconnected';
+
+		this.init();
+		this.connect();
+	}
+
+	private static validateProviderUrl(providerUrl: string): boolean {
+		return typeof providerUrl === 'string' ? /^ws(s)?:\/\//i.test(providerUrl) : false;
+	}
+
+	public getStatus(): Web3BaseProviderStatus {
+		return this.providerStatus;
+	}
+
+	/* eslint-disable class-methods-use-this */
+	public supportsSubscriptions(): boolean {
+		return true;
+	}
+
+	public on<T = JsonRpcResult>(
+		type: 'message' | string,
+		callback: Web3BaseProviderCallback<T>,
+	): void {
+		this.wsEventEmitter.on(type, callback);
+	}
+
+	public once<T = JsonRpcResult>(type: string, callback: Web3BaseProviderCallback<T>): void {
+		this.wsEventEmitter.once(type, callback);
+	}
+
+	public removeListener(type: string, callback: Web3BaseProviderCallback): void {
+		this.wsEventEmitter.removeListener(type, callback);
+	}
+
+	public connect(): void {
+		try {
+			this.providerStatus = 'connecting';
+
+			this.webSocketConnection = new W3WS(
+				this.clientUrl,
+				this.wsProviderOptions?.protocol,
+				undefined,
+				this.wsProviderOptions?.headers,
+				this.wsProviderOptions?.requestOptions,
+				this.wsProviderOptions?.clientConfig,
+			);
+
+			this.addSocketListeners();
+		} catch (e) {
+			this.providerStatus = 'disconnected';
+			throw new InvalidConnectionError(this.clientUrl);
+		}
+	}
+
+	public disconnect(code: number, reason: string): void {
+		this.removeSocketListeners();
+		this.webSocketConnection?.close(code || 1000, reason);
+		this.providerStatus = 'disconnected';
+	}
+
+	public reset(): void {
+		this.sentQueue.clear();
+		this.requestQueue.clear();
+
+		this.init();
+		this.removeSocketListeners();
+		this.addSocketListeners();
+	}
+
+	public async request<T = JsonRpcResponse, T2 = unknown[]>(
+		request: JsonRpcPayload<T2>,
+	): Promise<T> {
+		if (this.webSocketConnection === undefined)
+			throw new Error('WebSocket connection is undefined');
+
+		if (request.id === undefined) throw new Error('Request Id not defined');
+
+		if (this.requestQueue.has(request.id) || this.sentQueue.has(request.id))
+			throw new Error(
+				'Duplicate request Id. Another request is already in Queue with same request Id.',
+			);
+
+		const { id } = request;
+
+		if (
+			this.webSocketConnection.readyState === this.webSocketConnection.CLOSED ||
+			this.webSocketConnection.readyState === this.webSocketConnection.CLOSING
+		) {
+			this.providerStatus = 'disconnected';
+
+			this.requestQueue.delete(id);
```

Comment:
Shouldn't the `id` not exist in `this.requestQueue` here? The[ following code](https://github.com/ChainSafe/web3.js/blob/d3bc30822e132ebc1a9d94862e834f99f8d94936/packages/web3-providers-ws/src/index.ts#L145) even suggests that:

```typescript
if (this.requestQueue.has(request.id) || this.sentQueue.has(request.id))
    throw new Error(
        'Duplicate request Id. Another request is already in Queue with same request Id.',
    );
```

Reply:
it was for id collision check, if user passes request with same id that is already in Queue. 
But clean approach will be that transport shouldn't manage id, it should be moved to request manage.